Font support with freetype

GDQR opened the pull request:
-Font Render
-Load with Freetype
-Set the font size.
-font color support
-Set correct color.
-Support for setting the maximum font texture cache size. -Texture loaded with 32bit
-New tutorial


Need #174 and #175 to work.

use this font for the example https://fonts.google.com/specimen/Roboto, it is licensed Apache version 2.0

https://github.com/h4570/tyra/assets/104105647/3867fc5f-c273-47a1-a085-58e0c9a9ce30

GDQR committed:
Font tutorial 11

-Font Render
-Load with Freetype
-Support for setting the maximum font texture cache size.
-Texture loaded with 32bit
-New tutorial

GDQR committed:
Update Makefile

GDQR commented:
I fix it later.

GDQR committed:
fixes with makefiles.

h4570 commented:
Cool! 

Thank you for the contributing. 
I will try to find some time this week and check all of the PRs

GDQR commented:
No problem, meanwhile I will be solving these errors.

Wolf3s commented:
Tutorial's 10 class `Font` is conflicting with font.hpp. The error can be found here: https://github.com/h4570/tyra/blob/master/tutorials/10-sprite-sheet/inc/font_sprite.hpp

GDQR commented:
yeah, I'm going to delete the font of tutorial 10 and change it to the new one I made.
this is how it looks now.
I will make some corrections to make it look better
![tutorial10](https://github.com/h4570/tyra/assets/104105647/e2aa6aa0-d391-4c31-84d4-cd329521759c)

Wolf3s commented:
> yeah, I'm going to delete the font of tutorial 10 and change it to the new one I made. this is how it looks now. I will make some corrections to make it look better ![tutorial10](https://private-user-images.githubusercontent.com/104105647/299059613-e2aa6aa0-d391-4c31-84d4-cd329521759c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYwMzQ2NjIsIm5iZiI6MTcwNjAzNDM2MiwicGF0aCI6Ii8xMDQxMDU2NDcvMjk5MDU5NjEzLWUyYWE2YWEwLWQzOTEtNGMzMS04NGQ0LWNkMzI5NTIxNzU5Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEyM1QxODI2MDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MjlkMDczOTJlYTUwNTAwYzcyNzE4OGM4N2IyMGQ4Y2I2YTQ3ZTM2ZGJhMTA4NGJiNjlkMGE5ZDlkY2JiOTEyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.HE43GDjzhtHgZEqn2bx-VpM9EXiF7KKViHdtdjunIbU)

Nice stuff.

GDQR commented:
thanks

GDQR committed:
tutorial 10 optimization and font function change

GDQR committed:
clang format

GDQR committed:
Update Makefile

GDQR committed:
Fix Codacy static code

GDQR committed:
const reference and check buffer

GDQR commented:
ok, idk how to fix these 2 minor issues, I need help ;-; , also tutorial 10 now also needs the Roboto-black font.

GDQR committed:
Fix Codacy Static Code Analysis

h4570 commented:
It looks really promising! I will do review in a sec...

h4570 commented on engine/src/engine.cpp:
```diff
@@ -42,6 +42,7 @@ void Engine::initAll(const bool& loadUsbDriver) {
   banner.show(&renderer);
   audio.init();
   pad.init();
+  fontInit(&renderer);
```
New font class should be more object oriented (like `font.init(...)`, etc.).
Please refactor it to make it look like the neighboring classes (`pad`, `audio`, etc.)
Please avoid to use `extern` in following hpp file - probably is not needed here.

GDQR replied:
I thought about doing that but I also thought that an exception could be made 😅, since it was something that is used a lot and it would be easier to access the function directly. well now I change it.

h4570 commented on engine/inc/font/font.hpp:
```diff
@@ -0,0 +1,60 @@
+extern void unloadFont(Font* font);
+extern void drawText(Font* font, std::string text, float x, float y,
+                     int fontSize, Color color);
+// extern void drawText(std::string text, float x, float y, int fontSize,
```
Please try to remove all commented/temporary code

h4570 reviewed (changes requested):
Please try to fetch latest `master` code and change number of the tutorial to 12.

h4570 commented:
> ok, idk how to fix these 2 minor issues, I need help ;-; , also tutorial 10 now also needs the Roboto-black font.

What minor issues did you encounter? Maybe I can help with smth.

GDQR commented:
> What minor issues did you encounter? Maybe I can help with smth.

Don't worry, they were some problems with the codacy static code analysis, I already solved them. but thanks anyway.

GDQR committed:
Merge branch 'h4570:master' into font-ttf

GDQR committed:
object oriented, color support and more

-font color support
-individual character loading support
-option to disable texture limits

GDQR committed:
minor fixes

GDQR commented:
I have to fix something else but this is how it looks.
Fonts used:
https://fonts.google.com/specimen/Roboto (bold)
Segoe UI Emoji
![fontTyra](https://github.com/h4570/tyra/assets/104105647/9f531392-bdc1-4315-8322-d16a8170879a)

GDQR committed:
fix color support

GDQR commented:
I've finished now.

Wellinator commented:
Awesome job!!!

matt-comm-link commented:
Is this still in the pipeline?

GDQR commented:
not for now, it currently works, but it would be nice to update/add some things

matt-comm-link commented:
Would you say it's stable enough to migrate a project to in the meantime or would it be best left alone until it's ready to be merged?

GDQR commented:
yeah it's stable, I was using it for my own project and I didn't have any problems.

Wellinator commented:
@GDQR how's it going?!
Is this feature finished?
I was planning to add font support for my game.

GDQR commented:
I was busy with other things, I didn't advance on this, there are things that I think can be improved, like the spacing between letters but currently it works, later I will make another PR that has fewer commits, i clarify that i don't know when I will do it.

Wolf3s commented:
?

Wolf3s commented:
@GDQR
